### PR TITLE
[wip] evaluating render-links

### PR DESCRIPTION
**Pull request by kbhawkey**
Evaluating Hugo Render Hooks (requires Hugo > 0.62.0).
Installed Hugo v0.68.3 extended and modified config.toml (support Goldmark).
Added some test links in page `content/zh/docs/concepts/architecture/nodes.md`.
Added basic _markup template.

### CODE REVIEW

**Comment by kbhawkey** [comment]
See related pull request, https://github.com/kubernetes/website/pull/19526.

File: content/zh/docs/concepts/architecture/nodes.md [hunk 1]
```diff
@@ -26,6 +26,10 @@ architecture design doc for more details.
 -->
 在 Kubernetes 中，节点（Node）是执行工作的机器，以前叫做 `minion`。根据你的集群环境，节点可以是一个虚拟机或者物理机器。每个节点都包含用于运行 [pods](/docs/concepts/workloads/pods/pod/) 的必要服务，并由主控组件管理。节点上的服务包括 [容器运行时](/docs/concepts/overview/components/#node-components)、kubelet 和 kube-proxy。查阅架构设计文档中 [Kubernetes 节点](https://git.k8s.io/community/contributors/design-proposals/architecture/architecture.md#the-kubernetes-node) 一节获取更多细节。
 
+[scheduled](/docs/concepts/scheduling/)
```
Comment (tengqm) [inline content/zh/docs/concepts/architecture/nodes.md]: You have to use '/zh/docs/concepts/scheduling/' to navigate to the localized version, or you need to add logics to the render-link.html code below to get this done automagically.

Comment (kbhawkey) [reply content/zh/docs/concepts/architecture/nodes.md]: The snippet of the generated HTML page source:
```
<h1>节点</h1>



<!--
A node is a worker machine in Kubernetes, previously known as a `minion`. A node
may be a VM or physical machine, depending on the cluster. Each node contains
the services necessary to run [pods](/docs/concepts/workloads/pods/pod/) and is managed by the master
components. The services on a node include the [container runtime](/docs/concepts/overview/components/#node-components), kubelet and kube-proxy. See
[The Kubernetes Node](https://git.k8s.io/community/contributors/design-proposals/architecture/architecture.md#the-kubernetes-node) section in the
architecture design doc for more details.
-->
<p>在 Kubernetes 中，节点（Node）是执行工作的机器，以前叫做 <code>minion</code>。根据你的集群环境，节点可以是一个虚拟机或者物理机器。每个节点都包含用于运行 

<a href="/zh/docs/concepts/workloads/pods/pod/">pods</a>
 的必要服务，并由主控组件管理。节点上的服务包括 

<a href="/zh/docs/concepts/overview/components/#node-components">容器运行时</a>
、kubelet 和 kube-proxy。查阅架构设计文档中 

<a href="https://git.k8s.io/community/contributors/design-proposals/architecture/architecture.md#the-kubernetes-node" target="_blank">Kubernetes 节点</a>
 一节获取更多细节。</p>
<p>

<a href="/zh/docs/concepts/scheduling/">scheduled</a>



<a href="/zh/docs/concepts/#kubernetes-control-plane">Kubernetes control plane</a>



<a href="/zh/docs/admin/kubelet-authentication-authorization/">Kubelet authentication and/or authorization</a>
</p>
```

Comment (tengqm) [reply content/zh/docs/concepts/architecture/nodes.md]: The preview I saw is still something like this:
```
<p><a href="/docs/concepts/scheduling/">scheduled</a>
--
  | <a href="/docs/concepts/#kubernetes-control-plane">Kubernetes control plane</a>
  | <a href="/docs/admin/kubelet-authentication-authorization/">Kubelet authentication and/or authorization</a></p>
```

Maybe I was looking at the wrong place?

Comment (kbhawkey) [reply content/zh/docs/concepts/architecture/nodes.md]: Sorry for the confusion. I am testing locally with the latest version of Hugo.
The site is not set up to test this version.

Comment (kbhawkey) [reply content/zh/docs/concepts/architecture/nodes.md]: Just updated netlify.toml to use latest version of Hugo.
You should still see some links with problems.

https://deploy-preview-19830--kubernetes-io-master-staging.netlify.com/zh/docs/concepts/architecture/nodes/

Comment (tengqm) [reply content/zh/docs/concepts/architecture/nodes.md]: Now I can see the '/zh/' prefix is automatically added. The limits I saw so far: 1. intra-page links; 2. no auto fallback to English version.

Comment (kbhawkey) [reply content/zh/docs/concepts/architecture/nodes.md]: Yes. I guess the intra-page links can be corrected in the template. Wish there was a more "standard" way (integrated into Hugo) of creating links for multi-language doc sites.

**Comment by kbhawkey** [comment]
With the addition of the render-link template, the links in the localized pages are converted.
However, in the case where there are links between translated and non-translated files, there are potential problems (or unexpected results).
From what I can tell, all localizations display a mix of English and translated files in the left hand toc.

When the page links are converted, the links on the page point to the `fr` version of files.
With some links, the `fr` localization previously mapped to the English version of a page (page not translated). With the render-link conversion, the link now returns a 404 and not the `en` version of the file.
See links under:
https://deploy-preview-19830--kubernetes-io-master-staging.netlify.com/fr/docs/concepts/workloads/pods/pod-overview/pods-et-controllers

Another example:
If this page, https://deploy-preview-19830--kubernetes-io-master-staging.netlify.com/zh/docs/tasks/configure-pod-container/assign-memory-resource/
contains a link to,
https://deploy-preview-19830--kubernetes-io-master-staging.netlify.com/docs/tasks/configure-pod-container/configure-gmsa/

After the render-link template is processed, the links in `assign-memory-resource` will not point to the English file `configure-gmsa`, but expects the `zh` version of that file (which does not exist).

This would be surprising if the title of the `en` version page appears in the left hand toc and readers expect to forward to this `en` page from within the `zh` localization?

**Comment by kbhawkey** [comment]
@gochist @remyleone , Do you have any thoughts on using Hugo Render hooks for handling intra-site links?

**Comment by remyleone** [comment]
@kbhawkey I have not started yet to use it. Is there any getting started guide for localization teams?

**Comment by kbhawkey** [comment]
> 
> 
> @kbhawkey I have not started yet to use it. Is there any getting started guide for localization teams?

Hi! This is a follow-up to PR #19526 (use `ref` shortcode to improve intra-site links).
Since the link changes affect the site, I thought it would be a good idea to ask for input from localization teams.
Here is the link to the Hugo Markdown Render Hooks, https://gohugo.io/getting-started/configuration-markup#markdown-render-hooks. In pr 19526, it was suggested to use the new render template to update the site links. This pr is evaluating a custom template. The render hooks option is only available in Hugo versions >= 0.62.0. I _assume_ the template applies to every Markdown page, but has some restrictions (Hugo: "It’s currently only possible to have one set of render hook templates, e.g. not per Type or Section ..."). Changing the links seems problematic when the site includes a mix of English pages and translated pages per localization.

**Comment by sftim** [comment]
Can I help with this work?

**Comment by kbhawkey** [comment]
> Can I help with this work?
Yes, @sftim. What are your thoughts on using the render-link template?

**Comment by sftim** [comment]
> &hellip;@sftim. What are your thoughts on using the render-link template?

It sounds good. I've only limited experience customizing Hugo; I can probably find some time to learn more about it.
What are the shortcomings of the approach so far / where are those listed? Also, is there a particular subtask you'd like me to pick up?

**Comment by markthink** [comment]
image


@kbhawkey   这涉及到公共模板的修改，影响面会比较大，需要我支持么？

**Pull request closed without merging**